### PR TITLE
Switch from RocksDB to Redis

### PR DESCRIPTION
**Pull request by mperham**
This rewrites the storage system to use Redis, which is:

* easier to introspect
* cleaner on the filesystem (one RDB file!)
* higher level
* pure Go, no more C++ dependency

There are still opportunities for improvement but everything appears to work under normal conditions.

### CODE REVIEW

**Comment by peterkeen** [comment]
Does Faktory run it’s own redis process locally or can it talk over the network?

**Comment by mperham** [comment]
Right now it forks Redis with its own managed config, so as to ensure performance and proper configuration.  How Faktory interacts with replicas is TBD but one of the key reasons for porting to Redis was access to replication and failover tooling.

**Comment by mperham** [comment]
It launches Redis here:

https://github.com/contribsys/faktory/blob/redis/storage/redis.go#L89-L104

**Comment by krainboltgreene** [comment]
What was the onus for this change?

**Comment by mperham** [comment]
> one of the key reasons for porting to Redis was access to replication and failover tooling

**Comment by krainboltgreene** [comment]
@mperham Haha, I'm so sorry, I don't know how I missed that. Thanks for replying, I love your work.

**Comment by TheCorp** [comment]
You had previously said RocksDB was chosen because, "Redis doesn't have the necessary primitives to implement efficient multi-queue reliability easily. It took me years and multiple tries before I got it working well."

Does this mean the new redis implementation is equivalent to what you have working with Sidekiq or is it materially different?

(also hey @krainboltgreene !)

**Comment by mperham** [comment]
@TheCorp All Sidekiq processes talk to Redis directly so I can't control access.  Faktory is a gatekeeper in front of a private Redis so it has more options for doing things that Redis can't do itself.  It might be that we can implement a reliable multi-queue operation that isn't too onerous since all pushes and pops go thru Faktory.

**Comment by cdrx** [comment]
🎉

**Comment by mperham** [comment]
👋 RocksDB

File: Dockerfile [hunk 1]
```diff
@@ -1,34 +1,26 @@
...
                   /
-RUN apk add --no-cache libstdc++ libgcc
+
 RUN mkdir -p /root/.faktory/db
 RUN mkdir -p /var/lib/faktory/db
 RUN mkdir -p /etc/faktory
 
 EXPOSE 7419 7420
-CMD ["/faktory", "-b", "0.0.0.0:7419", "-e", "development"]
+CMD ["/faktory", "-w" "0.0.0.0:7420", "-b", "0.0.0.0:7419", "-e", "development"]
```
Comment (jbielick) [inline Dockerfile]: This looks like it's missing a comma `"-w", "0.0.0`

**Comment by dannykopping** [comment]
Will it still be possible to introspect on the embedded redis?

**Comment by mperham** [comment]
@dannykopping To open a redis-cli:

```
        docker run --rm -it \
                -v faktory-data:/var/lib/faktory \
                contribsys/faktory:$(VERSION) /usr/bin/redis-cli -s /var/lib/faktory/db/redis.sock
```

**Comment by eborden** [comment]
Is there a migration path from 0.8.0 to 0.9.0?

**Comment by mperham** [comment]
@eborden There's no migration path except to run them both in parallel until the old jobs are gone.  I've versioned Faktory as pre-1.0 software with the implication being that the project is not stable yet.  That said, I'm far happier with Redis now; I hope this means I won't need to make any further major changes without some sort of migration path.

**Pull request merged**